gdk/common/recipe_file/CaseInsensitiveRecipeFile.py

```diff
@@ -0,0 +1,92 @@
+import json
+import yaml
+from pathlib import Path
+from requests.structures import CaseInsensitiveDict
+
+
+class CaseInsensitiveRecipeFile:
+    def write(self, file_path: Path, content: CaseInsensitiveDict) -> None:
+        """
+        Writes CaseInsensitiveDict contents to a JSON or a YAML file based on the file path.
+        """
+        self._write(file_path, self._to_dict(content))
+
+    def read(self, file_path: Path) -> CaseInsensitiveDict:
+        """
+        Reads a JSON or a YAMl file contents as a CaseInsensitiveDict and returns it.
+        """
+        return self._from_dict(self._read(file_path))
+
+    def _write(self, file_path, content):
+        if self._is_json(file_path):
+            self._write_to_json(file_path, content)
+        elif self._is_yaml(file_path):
+            self._write_to_yaml(file_path, content)
+        else:
+            raise Exception(f"Invalid recipe file : {file_path}. Recipe file must be in json or yaml format.")
+
+    def _read(self, file_path):
```

Comment:
We could simplify this class a bit to reduce the cyclomatic complexity and the amount of indirection while making it a bit more declarative by doing something along this lines

```
read = {
   ".json":  json.loads,
   ".yaml":  yaml.safe_load,
  ".yml":  yaml.safe_load,
}

write = {
    ".yaml": yaml.safe_dump,
    ".yml":  yaml.safe_dump,
    ".json":  json.dump
}

def read(file_path: Path) -> CaseInsensitiveDict:
    _read = read.get(file_path.suffix)

    if _read is None:
          raise Exception(f"Invalid recipe file : {file_path}. Recipe file must be in json or yaml format.")

     with open(file_path, "r", encoding="utf-8") as f:
            return self._from_dict(_read(f.read()))


def write(content: file_path: CaseInsensitiveDict) -> None:
    _write = write.get(file_path.suffix)

    if _write in None:
           raise Exception(f"Invalid recipe file : {file_path}. Recipe file must be in json or yaml format.")

     with open(file_path, "w", encoding="utf-8") as f:
            _write(content, f)
```

Reply:
We're using some additional arguments for both `json.dump` and `yaml.safe_dump`. They also take the file argument in different order.